mhs/common/mhs_common/workflow/sync_async.py

```diff
@@ -47,18 +52,30 @@ def __init__(self,
         await work_description.set_status(wd.MessageStatus.INBOUND_RESPONSE_RECEIVED)
 
         try:
-            logger.info('002', 'Attempting to add inbound message to sync-async store')
-            await self.sync_async_store.add(message_id, {
-                CORRELATION_ID: correlation_id,
-                MESSAGE_DATA: payload
-            })
+            await self._add_to_sync_async_store(message_id, {CORRELATION_ID: correlation_id, MESSAGE_DATA: payload})
+            logger.info('004', 'Placed message onto inbound queue successfully')
+            await work_description.set_status(wd.MessageStatus.INBOUND_SYNC_ASYNC_MESSAGE_STORED)
         except Exception as e:
-            logger.warning('003', 'Failed to put message into sync-async store due to {Exception}', {'Exception': e})
+            logger.error('005', 'Failed to write to sync-async store')
             await work_description.set_status(wd.MessageStatus.INBOUND_SYNC_ASYNC_MESSAGE_FAILED_TO_BE_STORED)
             raise e
 
-        logger.info('004', 'Placed message onto inbound queue successfully')
-        await work_description.set_status(wd.MessageStatus.INBOUND_SYNC_ASYNC_MESSAGE_STORED)
+    async def _add_to_sync_async_store(self, key, data):
+        logger.info('002', 'Attempting to add inbound message to sync-async store')
+        retry = self.sync_async_store_max_retries
+        while True:
+            try:
+                await self.sync_async_store.add(key, data)
+                logger.info('003', 'Successfully updated state store')
+                break
+            except Exception as e:
+                logger.warning('021', 'Exception raised while adding to sync-async store {exception} {retry}',
+                               {'exception': e, 'retry': retry})
+                if retry == 0:
+                    logger.error('022', 'Final retry has been attempted for adding message to sync async store')
+                    raise e
+                retry -= 1
```

Comment:
Because this line is before the if statement, there's an out-by-one error, where if eg the number of retries is set to 1, then 2 attempts are made. Actually, you could argue that this is correct, but in any case, I think it's different to the retrying in the outbound transmission and the one I was doing in the inbound async-express workflow, and I think they should all be consistent.